Fracture C integral for creep crack growth analysis

recuero opened the pull request:
This PR implements the C(t) integral in the `TensorMechanics` module leveraging the existing code design that implements the J-integral using the domain integral. 

Code design changes are minimal. `JIntegral` and `DomainIntegralAction` now have an additional option to compute the modified J integral or C(t) integral.

Refs. #15517

Used for verification: Yoon, Park, Saxena, 2003, "Creep crack growth analysis of elliptic surface cracks in pressure vessels", Int. J. Press. Ves. Pip., 465-479

recuero committed:
Add strain energy rate density Refs #15517

recuero committed:
Eshelby tensor additions, JIntegral, and CrackFrontDefinition

recuero committed:
Incorporate C integral to DomainIntegral action and documentation

recuero committed:
Add J-integral tests

recuero committed:
Add and update domain integral tests

recuero committed:
Add mesh and journal files

recuero committed:
Add missing mesh and convert XFEM/SolidMechanics tests

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/555976/) on cfcb52c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15518/site)

This comment will be updated on new commits.

recuero commented:
For a refined mesh, and using expressions (13)-(14) by Yagawa of Yoon et al (2003), below is a comparison with MOOSE's results. There is a factor of 2 for n=4 and n=5 of the creep model, but this is consistent with the data presented in Fig. 6 of the same paper. Overall, MOOSE results appear to be in the ballpark. 

![image](https://user-images.githubusercontent.com/14300313/86370164-99a4de80-bc3c-11ea-89c2-49e2ea17ad09.png)

recuero commented:
@bwspenc Grizzly tests are failing because I am removing `convert_J_to_K` in this PR. If, after review, this change is okay with you, we'd need to update the input files of those few tests.

bwspenc commented on modules/tensor_mechanics/doc/content/modules/tensor_mechanics/FractureIntegrals.md:
```diff
@@ -106,3 +106,22 @@ The $T$-stress is the first second-order parameter in Williams' expansion of str
+
+## C-Integral
+
+The C(t) integral can be described as a path integral that is obtained by replacing displacements and strains with displacement rates and strain rates
```
The C(t) integral is a fracture integral that is obtained in essentially the same way as the $J$-integral described above, but by replacing displacements and strains with displacement rates and strain rates:

recuero replied:
Should be done

bwspenc commented on modules/tensor_mechanics/include/actions/DomainIntegralAction.h:
```diff
@@ -143,4 +151,6 @@ class DomainIntegralAction : public Action
   bool _solid_mechanics;
   /// Whether the constitutive models for the mechanics calculations use an incremental form
   bool _incremental;
+  /// Exponent on the effective stress
```
Add 'used by the C integral' to this comment.

recuero replied:
Should be done

bwspenc commented on modules/tensor_mechanics/doc/content/modules/tensor_mechanics/FractureIntegrals.md:
```diff
@@ -106,3 +106,22 @@ The $T$-stress is the first second-order parameter in Williams' expansion of str
+
+### Usage
+
+To compute $C$-integrals, the [DomainIntegral Action](/DomainIntegralAction.md) should be used to set up all of the required objects. In particular, two additional inputs are required `output_type = C` and `n_exponent`. The former refers to the type of integral requested ('C') and the latter is the matching exponent of the creep deformation material model.
```
We probably ought to say something about how `n_exponent` comes into the equation. It sounds like it makes some assumptions about the creep behavior following a power law model?  I guess I need to look into the papers a little bit to understand how that works.

recuero replied:
Yes, there are some assumptions on steady state creep around the crack tip that, together with experimental data, arrives at the factor \frac{n}{n+1} that modifies the strain energy rate density. Added some extra lines.

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/StrainEnergyRateDensity.md:
```diff
@@ -0,0 +1,33 @@
+  \label{eqn:sed_integral_def}
+  \dot{u} = \sigma : \dot{\epsilon}
+\end{equation}
+where $\sigma$ is the stress and $\epsilon$ is the strain rate. In the
```
I guess `n_exponent` actually comes in here, right?

recuero replied:
Yes. I left the formula without the exponent influence since, in theory, it's just a particular case, but explained it in a comment below.

bwspenc commented on modules/tensor_mechanics/src/materials/StrainEnergyRateDensity.C:
```diff
@@ -0,0 +1,65 @@
+//* This file is part of the MOOSE framework
```
So it looks like you elected to add an option to EshelbyTensor for the rate form, but add another separate class for the rate form of the strain energy density. I guess I don't have very strong feelings either, way, but we should talk a little about what motivated you to implement it the way you did.

recuero replied:
Adding an option to `EshelbyTensor` allows for some computation/variable retrieval code reuse, whereas the strain energy density rate is computed in one line of code, which is new. Also, it may be that the `EshelbyTensor` will only be used within the context of the `DomainIntegralAction` so adding an option to it may not mess things up too much. We could eventually want to use the `StrainEnergyRateDensity` for a different application. This second thought is more speculative though.

bwspenc replied:
Makes sense.

bwspenc reviewed (changes requested):
Overall, looks good!  A few details to work out, but I think this is pretty close.

bwspenc commented on modules/tensor_mechanics/src/actions/DomainIntegralAction.C:
```diff
@@ -56,8 +56,12 @@ DomainIntegralAction::validParams()
-  params.addParam<bool>(
-      "convert_J_to_K", false, "Convert J-integral to stress intensity factor K.");
+  MooseEnum output_type("J K C", "J");
+  params.addParam<MooseEnum>("output_type",
```
One more thing -- we should keep the old option as a deprecated parameter, at least long enough to fix up the Grizzly tests so we don't break it with this PR.

recuero replied:
Should work now.

bwspenc replied:
I think we don't even want `output_type` as a parameter here.

recuero committed:
Improve documentation and deprecated domain integral's conevrt_J_to_K parameter

bwspenc commented on modules/tensor_mechanics/src/actions/DomainIntegralAction.C:
```diff
@@ -118,7 +129,7 @@ DomainIntegralAction::DomainIntegralAction(const InputParameters & params)
     _treat_as_2d(getParam<bool>("2d")),
     _axis_2d(getParam<unsigned int>("axis_2d")),
-    _convert_J_to_K(false),
+    _output_type(getParam<MooseEnum>("output_type")),
```
I just realized something. I guess this is only applicable if you're computing J. It's also kind of weird that if you want C, you have to ask for the J integral and then set output_type if you want C.  What if we got rid of this option, and instead added 'C_integral' and 'K_from_J_integral' as options for 'integrals'?

recuero replied:
Should be done now. `CIntegral`, `JIntegral`, and `KFromJIntegral` are explicitly exclusive in the action with the latest changes.

bwspenc reviewed (changes requested):
A few really minor comments, and one thing I think we should change with the inputs. Otherwise, it's really close. I'm happy to discuss the input parameter options.

recuero committed:
Remove output_type from domain integral and augment 'integrals' with CIntegral and KFromJIntegral

recuero commented on modules/tensor_mechanics/src/materials/EshelbyTensor.C:
```diff
@@ -19,26 +19,36 @@ EshelbyTensor::validParams()
                                "Optional parameter that allows the user to define "
                                "multiple mechanics material systems on the same "
                                "block, i.e. for multiple phases");
+  params.addParam<bool>("compute_dissipation",
```
It is declared conditionally now. I have changed `Eshelby_tensor_rate` to `Eshelby_tensor_dissipation` to be more consistent. Also, I think "dissipation" is more meaningful than "rate" in this context.

recuero commented on modules/tensor_mechanics/doc/content/source/materials/StrainEnergyRateDensity.md:
```diff
@@ -0,0 +1,32 @@
+  \label{eqn:sed_integral_def}
+  \dot{u} = \int \boldsymbol{\sigma} : \textrm{d}\dot{\boldsymbol{\epsilon}}
+\end{equation}
+where $\boldsymbol{\sigma}$ is the stress tensor and $\dot{\boldsymbol{\epsilon}}$ is the strain rate. This expression is multiplied by $\frac{n}{n+1}$ when the input argument `n_exponent` is supplied. This factor decreases the strain energy rate density to better capture the strain rate field around a crack under steady-state creep growth. 
```
Should be added now.

recuero commented on modules/tensor_mechanics/include/materials/StrainEnergyRateDensity.h:
```diff
@@ -0,0 +1,45 @@
+#include "RankTwoTensorForward.h"
+
+/**
+ * StrainEnergyRateDensity calculates the strain energy density.
```
Done

recuero committed:
Remove debugging statement update module tests

bwspenc commented on modules/tensor_mechanics/src/actions/DomainIntegralAction.C:
```diff
@@ -93,6 +98,13 @@ DomainIntegralAction::validParams()
+                                  false,
+                                  "Convert J-integral to stress intensity factor K.",
+                                  "This input parameter is deprecated and will be removed soon. "
+                                  "Use 'output_type = K' to request output of the "
```
I think you'd want to use 'integrals = KFromJIntegral' instead, wouldn't you?

bwspenc commented on modules/tensor_mechanics/include/actions/DomainIntegralAction.h:
```diff
@@ -49,6 +51,13 @@ class DomainIntegralAction : public Action
     TOPOLOGY
   };
 
+  /// Enum for output type
```
I don't think we need this anymore.

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/StrainEnergyRateDensity.md:
```diff
@@ -0,0 +1,32 @@
+\end{equation}
+where $\boldsymbol{\sigma}$ is the stress tensor and $\dot{\boldsymbol{\epsilon}}$ is the strain rate. This expression is multiplied by $\frac{n}{n+1}$ when the input argument `n_exponent` is supplied. This factor decreases the strain energy rate density to better capture the strain rate field around a crack under steady-state creep growth. This factor is primarily used to compute the C(t) integr ...
+
+In the tensor mechanics module we define the strain rate as the sum of the
```
I'd slightly re-word this (because I don't think this is applicable only to tensor_mechanics):

The strain rate here is the sum of the elastic and inelastic (e.g. plastic, creep) strain rates.

bwspenc reviewed (changes requested):
Some really minor cleanup.

recuero committed:
Address reviewer's comments

recuero commented:
@bwspenc Instead of `output_type`, now `JIntegral` uses an enum `_integral` which can be any of the three types of output from the postprocessor: `JIntegral`, `CIntegral`, and `KFromJIntegral`. I think this is a bit more consistent compared to how it was before, when the `output_type` was still a thing in `JIntegral`. Your comments should be addressed now.

recuero committed:
Use enum class in JIntegral

bwspenc commented on modules/tensor_mechanics/include/materials/RadialReturnCreepStressUpdateBase.h:
```diff
@@ -25,8 +25,12 @@ class RadialReturnCreepStressUpdateBase : public RadialReturnStressUpdate
   virtual void propagateQpStatefulProperties() override;
-  virtual void computeStressFinalize(const RankTwoTensor & plastic_strain_increment) override;
-
+  virtual void computeStressFinalizeMain(const RankTwoTensor & plastic_strain_increment,
```
This will be the new version, right? You should just overload the name of the other one and name it `computeStressFinalize`, and then we'll get rid of the old one once we switch over the apps.

recuero replied:
Doing that causes an overload hiding error because all the derived classes should implement all or none of the methods with the same name (this appears to be in the language to avoid unintended consequences which should not apply here). Since we do not want to make changes to BISON, I believe changing the name is the option we have.

recuero committed:
Create an is_ad StrainEnergyRateDensity material

recuero committed:
First test

recuero committed:
Adapt tests to new implementation.

Tests need to use power law creep. New gold files are generated.

recuero committed:
Documentation fixes

recuero committed:
Loosen stress tolerances

bwspenc commented on modules/tensor_mechanics/include/materials/RadialReturnCreepStressUpdateBase.h:
```diff
@@ -22,6 +22,19 @@ class RadialReturnCreepStressUpdateBase : public RadialReturnStressUpdate
 
+  /**
+   * This class computes strain energy rate density for StrainEnergyRateDensity.
+   * The computed material property is needed for the fracture C(t) integral.
```
Use the standard doxygen format for methods:
```
/**
 * Compute the strain energy rate density for this inelastic model for the current step.
 * @param strain_energy_rate_density The computed strain energy rate density
 * @param stress The stress tensor at the end of the step
 * @param strain_rate The strain rate at the end of the step
 */
```
No need to include this for the derived class because it will be marked `override`, by the way.

bwspenc replied:
Also, seems like it would be better if this method returned a Real. Is there a reason to not do that?

recuero replied:
Switched to non-`void`

bwspenc commented on modules/tensor_mechanics/src/materials/ADPowerLawCreepStressUpdate.C:
```diff
@@ -78,3 +78,18 @@ ADPowerLawCreepStressUpdate::computeDerivative(const ADReal & effective_trial_st
+    const ADMaterialProperty<RankTwoTensor> & stress,
+    const ADMaterialProperty<RankTwoTensor> & strain_rate)
+{
+  if (_n_exponent < 1)
```
Does the exponent always need to be >1? In that case, maybe it should be a range-checked parameter.

recuero replied:
There is one test in `solid_mechanics` where the exponent is identically one: `./solid_mechanics/test/tests/combined_creep_plasticity/plasticity_only_combined_class_sm1.i`

But probably, for power law creep, it should be strictly greater than one.

recuero replied:
There is one test with n exponent identically one: `./solid_mechanics/test/tests/combined_creep_plasticity/plasticity_only_combined_class_sm1.i`

But it looks like power law creep requires exponent greater than one. Here, the intention should be avoiding division by zero. If the range is added in the material it would prevent the user from using it if it wants to experiment with exponent unity.

bwspenc reviewed (changes requested):
This looks really good! Just a couple of small comments.

recuero committed:
Address reviewer's comments II

recuero committed:
Address reviewers' comments III

Create templated type GenericRadialReturnCreepStressUpdateBase to avoid separate consideration of AD and non-AD classes through explicit casting.
